api/API_VERSIONING.md

```diff
@@ -0,0 +1,203 @@
+experience a backward compatible break on a change. Specifically:
+
+* Fields should not be renumbered or have their types changed. This is standard proto development
+  procedure.
+
+* Renaming of fields or package namespaces for a proto must not occur. This is inherently dangerous,
+  since:
+  * Fields renames break wire compatibility. This is stricter than standard proto development
+    procedure in the sense that it does not break binary wire format. However, it **does** break
+    loading of YAML/JSON into protos as well as text protos. Since we consider YAML/JSON to be first
+    class inputs, we must not change field names.
+
+  * For service definitions, the gRPC endpoint URL is inferred from package namespace, so this will
+    break client/server communication.
+
+  * For a message embedded in an `Any` object, the type URL, which the package namespace is a part
+    of, may be used by Envoy or other API consuming code. Currently, this applies to the top-level
+    resources embedded in `DiscoveryResponse` objects, e.g. `Cluster`, `Listener`, etc.
+
+  * Consuming code will break and require source code changes to match the API changes.
+
+* Some other changes are considered breaking for Envoy APIs that are usually considered safe in
+  terms of protobuf wire compatibility:
+  * Upgrading a singleton field to a repeated, e.g. `uint32 foo = 1;` to `repeated uint32 foo = 1`.
+    This changes the JSON wire representation and hence is considered a breaking change.
+  
+  * Wrapping an existing field with `oneof`. This has no protobuf or JSON/YAML wire implications,
+    but is disruptive to various consuming stubs in languages such as Go, creating unnecessary
+    churn.
+
+  * Increasing the strictness of
+    [protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate) annotations. Exceptions
+    may be granted for scenarios in which these stricter conditions model behavior already implied
+    structurally or by documentation.
+
+The exception to the above policy is for API versions tagged `vNalpha`. Within an alpha major
+version, arbitrary breaking changes are allowed. Also, within a `vN` major version, messages
+annotated as draft or experimental are exempt from this policy. The following annotations may be
+applied to a protobuf file, message, enum, field or enum value:
+* `[#proto-status: experimental]` indicating that backwards incompatible changes are likely and
+  major restructuring is anticipated.
+* `[#proto-status: draft]` indicating that backwards incompatible changes are possible but the
+  entity is near finalized and some implementation exists.
+* `[#not-implemented-hide:]` indicating that the entity is not implemented in Envoy and the entity
+  should be hidden from the Envoy documentation. No backwards incompatible changes are allowed.
+* `[#not-implemented-warn:]` indicating that the entity is not implemented in Envoy and the entity
```

Comment:
Is this used?

Reply:
In only a few places. I think we should eliminate it and converge with `not-implemented-hide`, will do a PR today to clean this up.